File: mytoolboxes/HeliCam/C4Utility/c4hdl/win64-x64/examples/Python/h8SurfSimple.py
##
# \file        h8SurfSimple.py
# \author      Silvan Murer
# \date        07 April 2021
# \copyright   Heliotis, 2021
# \version     1.0.2
#
# \brief       Simple Python example based on C4Hdl library for heliInspect H8
import os
import sys
import platform
import numpy as np
import matplotlib.pyplot as plt

if sys.version_info < (2, 7, 0):
    raise RuntimeError("Python 2.7 or later required")

if sys.platform == "win32":
    c4utilityRoot = os.environ['C4UTILITY_ROOT']
    libPath = c4utilityRoot + r'c4hdl'
    if (platform.architecture()[0] == "32bit"):
        libPath = libPath + r'\win32-i86'
    else:
        libPath = libPath + r'\win64-x64'
    if sys.version_info < (3, 0, 0):
        libPath = libPath + r'\py2.7\bin'
    else:
        libPath = libPath + r'\python\bin'
    sys.path.insert(0, libPath)
elif sys.platform == "linux":
    c4utilityRoot = os.environ['C4UTILITY_ROOT']
    libPath = c4utilityRoot + r'/c4hdl'
    if (platform.architecture()[0] == "32bit"):
        libPath = libPath + r'/linux32-i86'
    else:
        libPath = libPath + r'/linux64-x64'
    if sys.version_info < (3, 0, 0):
        libPath = libPath + r'/py2.7/bin'
    else:
        libPath = libPath + r'/python/bin'
    sys.path.insert(0, libPath)
else:
    # Only windows example is available
    raise RuntimeError('Not implemented now, only a windows example is available')

import C4HdlPY as c4hdl


##
# Scan for available interfaces and print interface list
# let the user select one and open the interface
# \param c4sys system object of C4Hdl library
# \return interface object of C4Hdl library
def selectInterface(c4sys):
    nofIf = c4sys.updateInterfaceList()
    print(str(nofIf) + " interfaces detected\n")
    for i in range(0, nofIf):
        ifName = c4sys.getInterfaceName(i)
        print(str(i+1) + ") " + ifName)
    selection_str = input("Select an interface (0=exit): ")
    selection_int = int(selection_str)
    if((selection_int <= 0) or (selection_int > nofIf)):
        exit('No interface selected - exit script')
    return c4sys.openInterface(selection_int - 1)


##
# Scan for available devices and print device list
# let the user select one and open the connection to the device
# \param c4if interface object of C4Hdl library
# \return device object of C4Hdl library
def selectDevice(c4if):
    nofDev = c4if.updateDeviceList()
    print(str(nofDev) + " devices detected\n")
    for i in range(0, nofDev):
        devName = c4if.getDeviceName(i)
        print(str(i+1) + ") " + devName)
    selection_str = input("Select a device (0=exit): ")
    selection_int = int(selection_str)
    return c4if.openDevice(selection_int - 1)


##
# simple configuration example of heliInspect H8 using internal motion control
# \param c4dev device object of C4Hdl library
def initializeDevice(c4dev):
    # Motion parameters in [mm] or [mm/s]
    scanPosition = -1.4
    scanRange = 0.5
    scanSpeed = 5.0
    generalSpeed = 10.0
    # exposure ration between 0.0 and 1.0 [1.0 = 100% exposure]
    exposureRatio = 1.0

    # enable required components
    c4dev.writeString("ComponentSelector", "Intensity")
    c4dev.writeInteger("ComponentEnable", 0)
    c4dev.writeString("ComponentSelector", "Range")
    c4dev.writeInteger("ComponentEnable", 1)
    c4dev.writeString("ComponentSelector", "Reflectance")
    c4dev.writeInteger("ComponentEnable", 1)
    c4dev.writeString("ComponentSelector", "Phase")
    c4dev.writeInteger("ComponentEnable", 0)

    # enable required chunk data (optional)
    c4dev.writeInteger("ChunkModeActive", 1)
    c4dev.writeString("ChunkSelector", "PartCount")
    c4dev.writeInteger("ChunkEnable", 1)
    c4dev.writeString("ChunkSelector", "PartType")
    c4dev.writeInteger("ChunkEnable", 1)

    # trigger configuration
    c4dev.writeString("TriggerSelector", "RecordingStart")
    c4dev.writeString("TriggerMode", "On")
    c4dev.writeString("TriggerSource", "Stage")
    c4dev.writeString("TriggerSelector", "AcquisitionStart")
    c4dev.writeString("TriggerMode", "Off")
    c4dev.writeString("TriggerSelector", "FrameStart")
    c4dev.writeString("TriggerMode", "On")
    c4dev.writeString("TriggerSource", "Software")
    # Hint: The TriggerSource also controls the TriggerSoftware feature.

    # encoder configuration
    c4dev.writeString("EncoderSelector", "Camera")
    c4dev.writeInteger("EncoderInverter", 1)

    # motion and position configuration
    c4dev.writeFloat("ScanPosition", scanPosition)
    c4dev.writeFloat("ScanRange", scanRange)
    c4dev.writeFloat("ScanSpeed", scanSpeed)
    c4dev.writeFloat("GeneralSpeed", generalSpeed)

    c4dev.writeString("ScanMode", "Down")
    c4dev.executeCommand("StageInit")

    # additional camera and processing configuration
    c4dev.writeString("Scan3dExtractionMethod", "AcceleratedCenterOfMassIQCorrection")
    c4dev.writeString("Scan3dScalingMethod", "zTags")
    c4dev.writeString("Scan3dDistanceUnit", "um")

    c4dev.writeFloat("TargetVerticalSpacing", 2.5)
    c4dev.writeFloat("ExposureRatio", exposureRatio)

    c4dev.writeString("FPNCorrection", "AverageLastFrames")
    c4dev.writeInteger("FPNCorrectionNFrames", 8)
    c4dev.writeInteger("ExtSimpMaxHWin", 7)

    # illumination control (D3)
    c4dev.writeString("LightControllerSelector", "LightController0")
    c4dev.writeString("LightControllerSource", "UserOutput0")
    c4dev.writeFloat("LightBrightness", 100.0)
    # illumination control (D2)
    # LineSelector = Line2
    # LineSource = UserOutput0
    # LineInverter = true
    # switch on the illumination (D2/D3)
    c4dev.writeString("UserOutputSelector", "UserOutput0")
    c4dev.writeInteger("UserOutputValue", 1)


print('********************************************************')
print('* Configure and acquire data in "3D surface mode"')
print('********************************************************')

c4sys = c4hdl.C4Handler()
c4if = selectInterface(c4sys)
c4dev = selectDevice(c4if)

try:
    print('initialize camera')
    initializeDevice(c4dev)
except Exception as error:
    print(error)
    exit('init failed')

scanPos = c4dev.readFloat("ScanPosition")
print("Ready for measurement. Measure on position: " + str(scanPos) + "mm")
selection_str = input("y=yes else=exit: ")
if (selection_str != "y"):
    c4dev.release()
    c4if.release()
    exit("User interrupt - exit")

print('start acquisition')
c4dev.startAcquisition(4)

# setup plot
gridsize = (1, 2)
fig = plt.figure()
axSurf = plt.subplot2grid(gridsize, (0, 0))
axAmp = plt.subplot2grid(gridsize, (0, 1))
fig.suptitle('Measurements')
axSurf.set_title('Surface')
imSurf = axSurf.imshow(np.ones([542, 512]))
axAmp.set_title('Amplitude')
imAmp = axAmp.imshow(np.ones([542, 512]))
plt.show(block=False)
plt.draw()
plt.pause(0.1)

# measurement loop (10 iterations)
for itr in range(0, 10):
    try:
        # trigger single measurement
        c4dev.writeString("TriggerSelector", "FrameStart")
        c4dev.executeCommand("TriggerSoftware")

        # acquire data
        c4buf = c4dev.getBuffer(10000)

        # look for float surface and amplitude component
        surfPart = -1
        ampPart = -1
        nofParts = c4buf.readInteger("ChunkPartCount")
        for i in range(0, nofParts):
            c4buf.writeInteger("ChunkPartSelector", i)
            partType = c4buf.readString("ChunkPartType")
            if(partType == "Surface"):
                surfPart = i
            elif(partType == "Amplitude"):
                ampPart = i

        # plot surface and amplitude component
        if (surfPart >= 0):
            rawSurf = c4buf.getDataPartFloat(surfPart).asNpArray()
            rawSurfDim = c4buf.getPartDimension(surfPart).asNpArray()
            surfWidth = rawSurfDim[0]
            surfHeight = rawSurfDim[1]
            surface = rawSurf.reshape(surfHeight, surfWidth)
            imSurf.set_data(surface)
            imSurf.set_clim(vmin=np.min(surface), vmax=np.max(surface))
            imSurf.set_extent((0, surfWidth, surfHeight, 0))
        if (ampPart >= 0):
            rawAmp = c4buf.getDataPartUint16(ampPart).asNpArray()
            rawAmpDim = c4buf.getPartDimension(ampPart).asNpArray()
            ampWidth = rawAmpDim[0]
            ampHeight = rawAmpDim[1]
            amplitude = rawAmp.reshape(ampHeight, ampWidth)
            imAmp.set_data(amplitude)
            imAmp.set_clim(vmin=np.min(amplitude), vmax=np.max(amplitude))
            imAmp.set_extent((0, ampWidth, ampHeight, 0))
        plt.draw()
        plt.pause(0.1)
        print("iteration " + str(itr) + " done!\n")

        # release buffer
        c4buf.release()
    except Exception as error:
        print(error)
        exit('acquisition error')

c4dev.stopAcquisition()
c4dev.release()
c4if.release()

File: mytoolboxes/HeliCam/C4Utility/c4hdl/win64-x64/examples/c/src/h8SurfSimple.c
/**
* \file        h8SurfSimpl.cpp
* \author      Silvan Murer
* \date        7 April 2021
* \copyright   Heliotis, 2021
*
* \brief Simple C example based on C4HdlC for heliInspect H8
*/
#include "C4HdlC.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * helper macro to check C function return values
 */
#define C_RETURN_CHECK(RETURN_VALUE, ERROR_MESSAGE)     \
if (RETURN_VALUE != C4HDL_ERR_SUCCESS) {   \
  printf(ERROR_MESSAGE);        \
  printf(C4Hdl_getLastError()); \
  return RETURN_VALUE;          \
}

#define C_RETURN_CHECK_AND_RELEASE(RETURN_VALUE, ERROR_MESSAGE)     \
if (RETURN_VALUE != C4HDL_ERR_SUCCESS) {   \
  printf(ERROR_MESSAGE);  \
  printf(C4Hdl_getLastError()); \
  if (c4dev != NULL) {    \
    C4Dev_release(c4dev); \
  } \
  if (c4if != NULL) {     \
    C4If_release(c4if);   \
  } \
  if (c4hdl != NULL) {    \
    C4Hdl_close(c4hdl);   \
  } \
  return RETURN_VALUE;    \
} 

/**
 * \brief  Scan for available interfaces and print interface list
 *         let the user select one and open the interface
 *
 * \param[in]     c4hdl  system object of C4HdlC library
 * \param[out]    selIf  pointer filled by function containing selected interface index
 *
 * \return C4HDL_ERROR error code
 */
C4HDL_ERROR selectInterface(C4_HANDLER c4hdl, int* selIf)
{
  C4HDL_ERROR err; /**< error handling */
  char cStr[1024];  /**< preallocated string buffer */
  size_t cStrSize = sizeof(cStr); /**< string buffer size */
  *selIf = -1;

  printf("Interfaces: \n");
  int64_t nofIf;
  err = C4Hdl_updateInterfaceList(c4hdl, &nofIf);
  C_RETURN_CHECK(err, "Failed to update interface list");
  if (nofIf == 0) {
    C_RETURN_CHECK(C4HDL_ERR_ERROR, "No interface detected!");
  }

  for (int64_t i = 0; i < nofIf; i++) {
    cStrSize = sizeof(cStr);
    err = C4Hdl_getInterfaceName(c4hdl, i, cStr, &cStrSize);
    C_RETURN_CHECK(err, "Failed to read interface name");
    printf("%lld: %s\n", (i + 1), cStr);
  }
  printf("Select a interfaces (exit with any other value)\n");
  /* wait for user input */
  char str[10];
  fgets(str, sizeof(str), stdin);
  int selection = atoi(str);
  selection--;

  if (selection < 0 || selection >= nofIf) {
    C_RETURN_CHECK(C4HDL_ERR_ERROR, "User Selection is not a valid interface!");
  }
  *selIf = selection;
  return C4HDL_ERR_SUCCESS;
}

/**
 * \brief  Scan for available devices and print device list
 *         let the user select one and open the connection to the device
 *
 * \param[in]   c4if  interface object of C4HdlC library
 * \param[out]  selDev  pointer filled by function containing selected device index
 *
 * \return C4HDL_ERROR error code
 */
C4HDL_ERROR selectDevice(C4_INTERFACE c4if, int* selDev)
{
  C4HDL_ERROR err; /**< error handling */
  char cStr[1024];  /**< preallocated string buffer */
  size_t cStrSize = sizeof(cStr); /**< string buffer size */

  *selDev = 0;

  printf("Devices: \n");
  int64_t nofDev;
  err = C4If_updateDeciveList(c4if, &nofDev);
  C_RETURN_CHECK(err, "Failed to update interface list");

  if (nofDev == 0) {
    C_RETURN_CHECK(C4HDL_ERR_ERROR, "No device detected!");
  }

  for (int64_t i = 0; i < nofDev; i++) {
    cStrSize = sizeof(cStr);
    err = C4If_getDeviceName(c4if, i, cStr, &cStrSize);
    C_RETURN_CHECK(err, "Failed to read device name");
    printf("%lld: %s\n", (i + 1), cStr);
  }
  printf("Select a device (exit with any other value)\n");
  /* wait for user input */
  char str[10];
  fgets(str, sizeof(str), stdin);
  int selection = atoi(str);
  selection--;

  if (selection < 0 || selection >= nofDev) {
    C_RETURN_CHECK(C4HDL_ERR_ERROR, "User Selection is not a valid device!");
  }
  *selDev = selection;
  return C4HDL_ERR_SUCCESS;
}

/**
 * \brief simple configuration example of heliInspect H8 using internal motion control
 *
 * \param[in]   c4dev  reference to a device object.
 */
C4HDL_ERROR initializeDevice(C4_DEVICE c4dev)
{
  C4HDL_ERROR err = 0; /**< simple error handling */
  // Motion parameters in [mm] or [mm/s]
  double scanPosition = -1.4;
  double scanRange = 0.5;
  double scanSpeed = 5.0;
  double generalSpeed = 10.0;
  // exposure ration between 0.0 and 1.0 [1.0 = 100% exposure]
  double exposureRatio = 1.0;

  // enable required components
  err += C4Dev_writeString(c4dev, "ComponentSelector", "Intensity");
  err += C4Dev_writeInteger(c4dev, "ComponentEnable", 0);
  err += C4Dev_writeString(c4dev, "ComponentSelector", "Range");
  err += C4Dev_writeInteger(c4dev, "ComponentEnable", 1);
  err += C4Dev_writeString(c4dev, "ComponentSelector", "Reflectance");
  err += C4Dev_writeInteger(c4dev, "ComponentEnable", 1);
  err += C4Dev_writeString(c4dev, "ComponentSelector", "Phase");
  err += C4Dev_writeInteger(c4dev, "ComponentEnable", 0);

  // enable required chunk data (optional)
  err += C4Dev_writeInteger(c4dev, "ChunkModeActive", 1);
  err += C4Dev_writeString(c4dev, "ChunkSelector", "PartCount");
  err += C4Dev_writeInteger(c4dev, "ChunkEnable", 1);
  err += C4Dev_writeString(c4dev, "ChunkSelector", "PartType");
  err += C4Dev_writeInteger(c4dev, "ChunkEnable", 1);

  // trigger configuration
  err += C4Dev_writeString(c4dev, "TriggerSelector", "RecordingStart");
  err += C4Dev_writeString(c4dev, "TriggerMode", "On");
  err += C4Dev_writeString(c4dev, "TriggerSource", "Stage");
  err += C4Dev_writeString(c4dev, "TriggerSelector", "AcquisitionStart");
  err += C4Dev_writeString(c4dev, "TriggerMode", "Off");
  err += C4Dev_writeString(c4dev, "TriggerSelector", "FrameStart");
  err += C4Dev_writeString(c4dev, "TriggerMode", "On");
  err += C4Dev_writeString(c4dev, "TriggerSource", "Software");
  // Hint: The TriggerSource also controls the TriggerSoftware feature.

  // encoder configuration
  err += C4Dev_writeString(c4dev, "EncoderSelector", "Camera");
  err += C4Dev_writeInteger(c4dev, "EncoderInverter", 1);

  // motion and position configuration
  err += C4Dev_writeFloat(c4dev, "ScanPosition", scanPosition);
  err += C4Dev_writeFloat(c4dev, "ScanRange", scanRange);
  err += C4Dev_writeFloat(c4dev, "ScanSpeed", scanSpeed);
  err += C4Dev_writeFloat(c4dev, "GeneralSpeed", generalSpeed);

  err += C4Dev_writeString(c4dev, "ScanMode", "Down");
  err += C4Dev_executeCommand(c4dev, "StageInit");

  // additional camera and processing configuration
  err += C4Dev_writeString(c4dev, "Scan3dExtractionMethod", "AcceleratedCenterOfMassIQCorrection");
  err += C4Dev_writeString(c4dev, "Scan3dScalingMethod", "zTags");
  err += C4Dev_writeString(c4dev, "Scan3dDistanceUnit", "um");

  err += C4Dev_writeFloat(c4dev, "TargetVerticalSpacing", 2.5);
  err += C4Dev_writeFloat(c4dev, "ExposureRatio", exposureRatio);

  err += C4Dev_writeString(c4dev, "FPNCorrection", "AverageLastFrames");
  err += C4Dev_writeInteger(c4dev, "FPNCorrectionNFrames", 8);
  err += C4Dev_writeInteger(c4dev, "ExtSimpMaxHWin", 7);

  // illumination control (D3)
  err += C4Dev_writeString(c4dev, "LightControllerSelector", "LightController0");
  err += C4Dev_writeString(c4dev, "LightControllerSource", "UserOutput0");
  err += C4Dev_writeFloat(c4dev, "LightBrightness", 100.0);
  // illumination control (D2)
  // LineSelector = Line2
  // LineSource = UserOutput0
  // LineInverter = true
  // switch on the illumination (D2/D3)
  err += C4Dev_writeString(c4dev, "UserOutputSelector", "UserOutput0");
  err += C4Dev_writeInteger(c4dev, "UserOutputValue", 1);

  return C4HDL_ERR_SUCCESS;
}


/**
 * \brief main function of the application
 *
 * \param[in]       argc     arguments are not used by this application
 * \param[in]       *argv[]  arguments are not used by this application
 *
 * \return application error code
 */
int main(int argc, char *argv[]) {
  printf("********************************************************\n");
  printf("* Configure and acquire data in '3D surface mode'\n");
  printf("********************************************************\n");

  // variable definitions 
  C4HDL_ERROR err; /**< error handling */
  char cStr[1024];  /**< preallocated string buffer */
  size_t cStrSize = sizeof(cStr); /**< string buffer size */
  
  C4_HANDLER c4hdl = NULL;  /**< system handle */
  C4_INTERFACE c4if = NULL; /**< interface handle */
  C4_DEVICE c4dev = NULL;   /**< device handle */
  C4_BUFFER c4buf0 = NULL;  /**< buffer handle */

  err = C4Hdl_open(&c4hdl);
  C_RETURN_CHECK_AND_RELEASE(err, "Failed to open C4Hdl");
  // Select and open Interface
  int ifNo = -1;
  err = selectInterface(c4hdl, &ifNo);
  C_RETURN_CHECK_AND_RELEASE(err, "Interface selection failed!");
  err = C4Hdl_openInterface(c4hdl, &c4if, ifNo);
  C_RETURN_CHECK_AND_RELEASE(err, "Failed to open interface");
  // Select and open Device
  int devNo = -1;
  err = selectDevice(c4if, &devNo);
  C_RETURN_CHECK_AND_RELEASE(err, "Device selection failed!");
  err = C4If_openDevice(c4if, &c4dev, devNo);
  C_RETURN_CHECK_AND_RELEASE(err, "Failed to open device");

  // Initialize device
  err = initializeDevice(c4dev);
  C_RETURN_CHECK_AND_RELEASE(err, "Device initialization failed!");

  // start acquisition
  err = C4Dev_startAcquisition(c4dev, 4);
  C_RETURN_CHECK_AND_RELEASE(err, "Could not switch to acquisition mode!");

  // measurement loop (10 iterations)
  for (int i = 0; i < 10; i++) {
    err = C4Dev_executeCommand(c4dev, "TriggerSoftware");
    C_RETURN_CHECK_AND_RELEASE(err, "Failed to execute software trigger!");

    err = C4Dev_getBuffer(c4dev, &c4buf0, 10000);
    C_RETURN_CHECK_AND_RELEASE(err, "Failed to acquire a buffer containing the measurement data!");

    err = C4Buf_release(c4buf0);
    C_RETURN_CHECK_AND_RELEASE(err, "Could not release buffer!");
    c4buf0 = NULL;

    printf("iteration %d done!\n", i);
  }

  err = C4Dev_stopAcquisition(c4dev);
  C_RETURN_CHECK_AND_RELEASE(err, "Could not stop acquisition mode!");

  err = C4Dev_release(c4dev);
  c4dev = NULL;
  C_RETURN_CHECK_AND_RELEASE(err, "Could not release device!");
  err = C4If_release(c4if);
  c4if = NULL;
  C_RETURN_CHECK_AND_RELEASE(err, "Could not release interface!");
  err = C4Hdl_close(c4hdl);
  c4hdl = NULL;
  C_RETURN_CHECK_AND_RELEASE(err, "Could not release handle!");
  return 0;
}
